src/hotspot/share/code/aotCodeCache.cpp

```diff
@@ -0,0 +1,1368 @@
+                                                entry_position, entry_size, name_offset, name_size,
+                                                blob_offset, has_oop_maps, blob.content_begin());
+  log_info(aot, codecache, stubs)("Wrote code blob '%s(id=%d)' to AOT Code Cache", name, id);
+  return true;
+}
+
+CodeBlob* AOTCodeCache::load_code_blob(AOTCodeEntry::Kind entry_kind, uint id, const char* name, int entry_offset_count, int* entry_offsets) {
+  AOTCodeCache* cache = open_for_use();
+  if (cache == nullptr) {
+    return nullptr;
+  }
+  assert(AOTCodeEntry::is_valid_entry_kind(entry_kind), "invalid entry_kind %d", entry_kind);
+
+  if ((entry_kind == AOTCodeEntry::Blob) && !cache->stub_caching()) {
+    return nullptr;
+  }
+  if ((entry_kind == AOTCodeEntry::Adapter) && !cache->adapter_caching()) {
+    return nullptr;
+  }
+  log_info(aot, codecache, stubs)("Reading blob '%s' from AOT Code Cache", name);
+
+  AOTCodeEntry* entry = cache->find_entry(entry_kind, id);
+  if (entry == nullptr) {
+    return nullptr;
+  }
+  AOTCodeReader reader(cache, entry);
+  return reader.compile_code_blob(name, entry_offset_count, entry_offsets);
+}
+
+CodeBlob* AOTCodeReader::compile_code_blob(const char* name, int entry_offset_count, int* entry_offsets) {
+  uint entry_position = _entry->offset();
+
+  // Read name
+  uint name_offset = entry_position + _entry->name_offset();
+  uint name_size = _entry->name_size(); // Includes '/0'
+  const char* stored_name = addr(name_offset);
+
+  if (strncmp(stored_name, name, (name_size - 1)) != 0) {
+    log_warning(aot, codecache, stubs)("Saved blob's name '%s' is different from the expected name '%s'",
+                                       stored_name, name);
+    ((AOTCodeCache*)_cache)->set_failed();
+    exit_vm_on_load_failure();
+    return nullptr;
+  }
+
+  // Read archived code blob
+  uint offset = entry_position + _entry->blob_offset();
+  CodeBlob* archived_blob = (CodeBlob*)addr(offset);
+  offset += archived_blob->size();
+
+  address reloc_data = (address)addr(offset);
+  offset += archived_blob->relocation_size();
+  set_read_position(offset);
+
+  ImmutableOopMapSet* oop_maps = nullptr;
+  if (_entry->has_oop_maps()) {
+    oop_maps = read_oop_map_set();
+  }
+
+  CodeBlob* code_blob = CodeBlob::create(archived_blob, stored_name, reloc_data, oop_maps);
+  if (code_blob == nullptr) { // no space left in CodeCache
+    return nullptr;
+  }
+
+  fix_relocations(code_blob);
+
+  // Read entries offsets
+  offset = read_position();
+  int stored_count = *(int*)addr(offset);
+  assert(stored_count == entry_offset_count, "entry offset count mismatch, count in AOT code cache=%d, expected=%d", stored_count, entry_offset_count);
+  offset += sizeof(int);
+  set_read_position(offset);
+  for (int i = 0; i < stored_count; i++) {
+    uint32_t off = *(uint32_t*)addr(offset);
+    offset += sizeof(uint32_t);
+    const char* entry_name = AdapterHandlerEntry::entry_name(i);
```

Comment:
This is Adapter specific code. But we use this API for blobs as well. I thing we should either remove it or refactor to make it more generic.

Reply:
I specialized `entry_name` for adapters otherwise "".